[1단계 - 출석 구현] 코기(장재현) 미션 제출합니다.

jaehyeon2650 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [코기 프롤로그](https://prolog.techcourse.co.kr/studylogs/3929)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
```규칙 4: 한 줄에 점을 하나만 찍는다.```

해당 규칙을 지키지 못한 코드가 존재한다.
객체에 메시지를 던져 특정 책임이 있는 객체에게 결과를 요구하는 설계를 하려고 노력했다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

코드에 대한 질문은 아니고 TDD에 대한 질문이 있습니다. 제가 TDD를 진행할 때 미리 도메인을 설계한 후에 테스트 작성을 하기 
시작했습니다. 그런데 TDD에 대해 강의를 들을 때 코치님은 최소한의 설계만 하고 바로 테스트 코드를 작성한다고 하셨습니다. 
TDD를 하면서 각 객체의 구성요소를 정하고 역할을 분배한다고 하셨습니다. 저는 여기서 최소한의 설계가 어느정도인지 감이 오지 않습니다.
이번에 TDD를 적용할 땐 미리 객체 구성요소를 설계 후 각 객체에 대해 TDD를 진행했는데 이렇게 하면 안되는건가요??
최소한의 설계가 어느 정도까지인지 궁금합니다!

Byesol committed:
docs(README) : 리드미 작성

Byesol committed:
feat(domain): History 도메인 생성

- 출석 결과 조회 기능 추가
- 출석 휴일 검사 기능 추가

Byesol committed:
feat(domain): Historyies 도메인 생성

- 출석 결과 카운팅 기능 추가
- 제적 레벨 판단 기능 추가

Byesol committed:
feat(domain) : History 기능 추가

- 출석 결과 추가

Byesol committed:
feat(domain): Crew 도메인 생성

- 이름 검증 로직 추가

Byesol committed:
feat(domain) : History 기능 추가

- 출석 결과 추가

Byesol committed:
feat(domain): Histories 기능 추가

- 출석 기록 수정 기능 추가

Byesol committed:
feat(domain): crew 기능 추가

- 출석 확인 기능 추가

Byesol committed:
feat(domain) : History 기능 추가

- Histories에 출석 기록 기능 추가

Byesol committed:
feat(domain) : History 기능 추가

- Histories에 출석 기록 기능 추가

Byesol committed:
feat(controller) : AttendanceController 추가

Byesol committed:
feat(util) : Convertor 추가

Byesol committed:
feat(view) : InputView 구현

- 이름 입력 기능 추가
- 등교 시간 입력 기능 추가

Byesol committed:
feat(view) : OutputView 구현

-  출석 기록 확인 기능 추가

Byesol committed:
feat(domain) : Histories 기능 추가

- 결석 기록 추가

jaehyeon2650 committed:
feat(domain): History 기능 추가

- 특정 날짜 이전 기록인지 확인하는 기능 추가

jaehyeon2650 committed:
fix(domain): 출석 결과 반환 로직 변경

- 출석 기록이 없는 경우 결석으로 처리

jaehyeon2650 committed:
feat(util): csv 파일 리더 생성

- 파일을 읽어 Map을 반환한다.

jaehyeon2650 committed:
feat(domain): Crews 도메인 생성

- 이전 출석 기록 조회 기능 추가
- 출석 기능 추가

jaehyeon2650 committed:
feat(domain): 오늘 출석 확인 기능 추가

jaehyeon2650 committed:
feat(domain): Crew 이전 출석 기록 조회 기능 추가

jaehyeon2650 committed:
feat(configure): 주입 기능 추가

jaehyeon2650 committed:
feat(util): 날짜 변경 로직 추가

- MM월 dd일 E요일 HH:mm 형태 반환
- MM월 dd일 E요일 형태 반환
- HH:mm 형태 반환
- 12월 시간으로 변경

jaehyeon2650 committed:
feat(view): 입력 기능 추가

- 수정 입력 기능 추가
- 메뉴 입력 기능 추가

jaehyeon2650 committed:
feat(view): 출력 기능 추가

- 수정 완료 출력 기능 추가

jaehyeon2650 committed:
feat(domain): Crews 기능 추가

- 유저의 출석 결과를 반환
- 유저의 출석 기록을 반환
- 유저의 출석 기록 수정

jaehyeon2650 committed:
feat(dto): dto 생성

jaehyeon2650 committed:
feat(domain): Histories 기능 추가

- 결석 관련 정보 저장 기능 추가

jaehyeon2650 committed:
feat(domain): History 정렬 기능 추가

jaehyeon2650 committed:
feat(view): 출석 기록 출력 기능 추가

jaehyeon2650 committed:
feat(utils): 결석 기록 포맷 추가

jaehyeon2650 committed:
feat: 출석 조회 기능 추가

- 컨트롤러 출석 조회 기능 추가
- Crew, Crews 출석 관련 정보 조회 기능 추가

jaehyeon2650 committed:
feat(dto): 재적 위험자 dto 생성

- 정렬 기능 추가

jaehyeon2650 committed:
feat(domain): Crews 제적 대상자 조회 기능 추가

jaehyeon2650 committed:
feat(view): 제적 위험자 출력 기능 추가

jaehyeon2650 committed:
feat(controller): 제적 위험자 출력 기능 추가

jaehyeon2650 committed:
feat(admin): 캠퍼스 운영 시간 예외 기능 추가

jaehyeon2650 committed:
feat(constants): 입력 상수 추가

jaehyeon2650 committed:
refactor(view): 스캐너 인스턴스를 한번만 생성하도록 변경

jaehyeon2650 committed:
refactor(view): OutputVIew 메서드 추출

- OutputVIew 메서드 추출
- LINE_SEPARATOR 추가
- 출력 형식 간소화

jaehyeon2650 committed:
feat(constants): 공휴일 enum 클래스 생성

Byesol committed:
refactor(test) : test 메서드 리팩토링

-  메서드 명 리팩토링
-  DisplayName

Byesol committed:
refactor(util) : CsvReader 메서드 분리

-  메서드 명 리팩토링
-  DisplayName
-  상수 분리

Byesol committed:
refactor(domain) : AttendanceResult enum 클래스 생성

Byesol committed:
refactor(domain) : AbsenceLevel enum 기능 추가

Byesol committed:
refactor(domain) : Crews 메소드 추출

Byesol committed:
refactor(view) : view 패키지 이름 수정

Byesol committed:
refactor(view) : view 패키지 이름 수정

Byesol committed:
refactor(domain) : 출석 수정 기능 버그 수정

Byesol committed:
refactor(contoller) : 메소드 분리

- 반복되는 로직 convertor 메소드에서 사용

Byesol committed:
refactor : 메소드 분리

Byesol committed:
refactor(test) : 테스트 리팩토링

- 케이스 추가
- 테스트 메소드명 변경

Byesol committed:
refactor(domain) : 메소드 리팩토링

- getter 함수 반환 타입 변경(AttendanceResult->string)
- 메소드명  getHistories-> getSortedHistories

Byesol committed:
refactor(docs) : 리드미 수정

jaehyeon2650 committed:
docs: pr 템플릿 작성

Rok93 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,10 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Configure configure = new Configure();
```
main 함수에서 AttendanceController 객체를 직접 생성해줘도 될 것 같은데, Configure를 두신 이유가 있을까요?

jaehyeon2650 replied:
프리코스에서 Main(Application을 실행하는 클래스)에서 애플리케이션을 실행하는 책임만을 가지도록 하고 객체를 생성하는 역할을 하는 클래스는 따로 만들라는 요구사항이 있었습니다. 
그 요구사항을 보니 저 역시도 Main클래스는 최소한의 객체 생성만 하고 실행만 하는 역할을 가지는게 맞다고 판단하여 Configure 클래스를 만들어 필요한 객체 생성과 파일 입출력을 통한 초기화 작업을 나눴습니다!

Rok93 replied:
Config에서 컨트롤러에 대한 셋팅을 진행하는 것은 좋다고 생각해요! 
다만 Config 내에서 도메인 객체 생성에 관여하는 것은 지양해야한다고 생각해요. (이유에 대해서는 Configure 클래스 쪽 코멘트에 남겨뒀어요! 😃)

Rok93 commented on src/main/java/configure/Configure.java:
```diff
@@ -0,0 +1,45 @@
+    private static AttendanceController attendanceController;
+    private static OutputVIew outputVIew;
+    private static InputView inputView;
+    private static Crews crews;
```
스프링의 Bean을 등록해주는 코드가 왠지 기억이 나는 방식이네요.  👀
스프링의 방식을 본따서 만드신거라면 여기서는 큰 의미가 없다고 생각해요. 

또한, 도메인 객체인 crews를 이런 config의 위치에서 만드는 것도 적절치 못하다고 생각해요.

jaehyeon2650 replied:
맞습니다! 스프링 bean 등록을 본따서 만들었습니다. 크루들의 정보를 파일 입출력을 통해 애플리케이션 실행동안 단 한번만 초기화하는데 이 특성이 싱글톤 특징과 비슷하여 이렇게 설계를 하였습니다.

> 도메인 객체인 crews를 이런 config의 위치에서 만드는 것도 적절치 못하다고 생각해요.

저는 configure이 컨트롤러 객체 생성과 파일 입출력을 통해 크루원 객체 초기화하는 역할을 가지도록 설계했습니다. 왜 config 위치에서 이런 행위를 하는게 적절치 못하는지 이해가되지않습니다! 로키가 이렇게 생각한 이유에 대해 자세히 설명해 주실 수 있을까요?🥹

Rok93 replied:
두 가지 관점이있는데요. 

1. Config라는 네이밍에서 설정 값만 다뤄야할 것 같은데 도메인 객체를 생성에 관여한다는 점

2. 사실 지금은 싱글 스레드로 동작할 것이기 때문에 동시에 여러 요청을 처리할 일이 없겠지만, 추후 조금 더 확장하여 이 출석 도메인 서비스가가 여러 사용자에게 동시에 호출되게되면 이 `Crews`가 전역 상태처럼 사용될 여지가 생길 수 있을 것 같아요.

jaehyeon2650 replied:
여러 사용자가 동시에 호출하는 상황까지는 생각하지 못했네요! 로키 말대로 이런 상황이 오면 객체가 공유 자원처럼 사용되면서, 내부 상태가 변하면 race condition 이 발생할 수 있을거 같네요!

로키 말대로 Configure에서 하지 않고 호출 시에 초기화하는 것이 나아보입니다! 로키 덕분에 관점도 늘어나네요! 감사합니다 👍

Rok93 commented on src/main/java/constants/SelectionOption.java:
```diff
@@ -0,0 +1,26 @@
+    GET_ATTENDANCE_HISTORY("3"),
+    CHECK_ABSENCE_USERS("4"),
+    QUIT("Q");
+    private final String option;
```
줄바꿈을 통해서 상태와 enum 요소를 구분하기 쉽게해주면 가독성이 좋아질 것 같아요. 😃 

```suggestion
    QUIT("Q");
    
    private final String option;
```

jaehyeon2650 replied:
지금보니 로키 말대로 공백 컨벤션을 잘 지키지 못한거 같네요! 좋은 지적 감사합니다.😊

Rok93 commented on src/main/java/constants/SelectionOption.java:
```diff
@@ -0,0 +1,26 @@
+
+    public static SelectionOption getSelectOption(String input){
+        return Arrays.stream(SelectionOption.values()).filter(select -> select.option.equals(input))
+                .findAny().orElseThrow(()->new IllegalArgumentException("[ERROR] 잘못된 입력입니다."));
```
한 줄에 하나의 메서드만 호출하도록 해주면 stream 코드의 가독성이 좋아질 것 같아요. 
추가적으로 코드 정렬도 잊지말고 꼭 부탁드릴게요. 😉

```suggestion
                .findAny()
                .orElseThrow(() ->new IllegalArgumentException("[ERROR] 잘못된 입력입니다."));
```

jaehyeon2650 replied:
로키 말대로 한 줄에 여러 메서드가 호출되어 가독성을 해치는거 같아요! 코드 정렬과 stream 코드 정렬 하도록 하겠습니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,94 @@
+        do {
+            answer = inputView.getMenu();
+            startMenu(answer);
+        } while (!(answer == SelectionOption.QUIT));
```
조건문을 `!=` 대신에 `!(...)`로 두신 이유가 있을까요? 
```suggestion
        } while (answer != SelectionOption.QUIT));
```

jaehyeon2650 replied:
앗..! 아무 이유 없습니다! 로키 코드처럼 `!=` 를 사용하는게 의미 전달이 확실할거 같아요!

Rok93 commented on src/main/java/domain/AbsenceLevel.java:
```diff
@@ -0,0 +1,36 @@
+        this.standard = standard;
+    }
+
+    public static AbsenceLevel findAbsenceLevel(int absentCount, int lateCount) {
```
이 함수의 입장에서는 순수하게 결석일수만 주입받아서 제적, 면담, 경고, 정상의 결과를 보여주는게 추후 지각 --> 결석 계산 방식이 바뀐다거나하는 변경에 영향이 적을 것 같아요. 🤔

jaehyeon2650 replied:
로키의 의견에 의문점이 있습니다! 만약 로키 말대로 결석일수만 주입 받아서 결과를 보여주도록 리팩토링을 하고 계산 방식이 바뀐다하면 결국 결석 계산(지각-> 결석으로 바꾸는 계산)은 Histories에서 할거같습니다. 
근데 만약 계산 방식이 바뀌면 Histories에서 변경을 가해야하는데 이러면 결국 변경이 일어나면 특정 객체 하나에서 일어나는건 똑같은거 아닌가요?? 지금 설계처럼했을 때도 계산 방식이 바뀌면 AbsenceLevel만 바꾸면 된다고 생각합니다. 
로키 말대로 리팩토링을 했을 때 장점을 정확히 이해하지 못했습니다!

Rok93 replied:
말씀하신 방향도 일리가 있다고 생각했는데, `출석일수, 지각일수, 결석일수`를 하나로 묶은 객체를 만든 뒤, 이 객체에게 지각일 수를 계산하는 역할을 위임해볼 수도 있겠다는 생각이 드네요. 😃 

물론 역할을 위임하는 것이 굳이 필요하지 않다면 지금처럼 AbsenceLevel에게 위임하는 방식도 나쁘지는 않다고 생각해요. 🤔

Rok93 commented on src/main/java/domain/AbsenceLevel.java:
```diff
@@ -0,0 +1,36 @@
+package domain;
+
+public enum AbsenceLevel {
```
결석 레벨이라는 네이밍이 직관적으로 어떤 것을 나타내는지 잘 이해되지 않는데요. 🤔 

요구사항을 다시 읽어보면서 조금 더 적절한 네이밍은 없을까요?

jaehyeon2650 replied:
처음 코드를 보는 입자에서 로키 말대로 한번에 이해되지 않을거 같아요!

현재 생각나는 이름으론 결석과 지각으로 인한 제재라고 생각되어 AbsencePenalty 정도가 생각나네요! 좀 더 고민을 해보고 이름을 변경하도록 하겠습니다!

Rok93 commented on src/main/java/domain/AttendanceResult.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.time.LocalDateTime;
+
+public enum AttendanceResult {
+    MONDAY_ABSENCE("결석", 13, 30),
+    MONDAY_LATE("지각", 13, 5),
+    ABSENCE("결석", 10, 30),
+    LATE("지각", 10, 5),
+    ATTENDANCE("출석", 12, 0);
```
`출석요일`을 enum 객체로 만들어서  `수업 시작 시간`을 상태로 가지도록 해보면 어떨까요? 🤔 

그렇다면 출석요일 객체가 `AttendanceResult`를 판단해주는 역할을 할 수 있을 것 같네요. 😃

jaehyeon2650 replied:
이 객체가 정말 애매하다고 생각되었는데 로키 의견대로하면 각 객체의 역할을 확실히 분리할 수 있을거 같습니다! 좋은 의견 정말 감사합니다. 한번 리팩토링 해보도록 하겠습니다.

Rok93 commented on src/main/java/domain/History.java:
```diff
@@ -0,0 +1,57 @@
+    private final static int END_TIME = 23;
+
+    public History(LocalDateTime attendanceTime) {
+        validateHistory(attendanceTime);
```
History라는 네이밍이 너무 포괄적이어서, 저는 여기서 왜 휴일에 대한 유효성 검증을 하는건지 잘 이해가 안됐었어요. 😢
`출석 이력` 이라고 네이밍 짓는다면 조금 더 직관적으로 이해할 수 있지 않을까요?

jaehyeon2650 replied:
History의 역할은 하루의 출석 상태를 관리하는 객체입니다. 그렇기에 휴일은 출석하지 않기에 이 객체에서 검증을 하는게 맞다고 판단하였습니다!
로키 의견대로 네이밍이 아쉽네요..! 다시 생각해보고 수정해보도록 하겠습니다! 좋은 의견 감사합니다 👍

Rok93 commented on src/main/java/domain/History.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private void validateOpeningHours(LocalDateTime attendanceTime) {
+        if (!(attendanceTime.getHour() == ABSENT_DEFAULT_HOUR && attendanceTime.getMinute() == ABSENT_DEFAULT_MINUTE)
```
이 부분도 `출석요일` 객체가 생긴다면 이 객체에게  메시지를 보내볼 수도 있을 것 같네요.

jaehyeon2650 replied:
오 그렇네요! 한번 리펙토링 해보겠습니다

Rok93 commented on src/main/java/domain/AttendanceResult.java:
```diff
@@ -0,0 +1,62 @@
+
+    private final String result;
+    private final int hour;
+    private final int mintues;
```
오타가 있는 것 같아요 👀

Rok93 replied:
추가적으로 이 부분 질문이있는데요! 
`LocalTime` 하나를 상태로 뒀어도 됐을 것 같은데, hour, minute를 분리해서 상태로 두신 이유가 있으신지 궁금합니다. 😃

jaehyeon2650 replied:
LocalTime의 존재를 몰랐습니다...ㅜ 지금까지 LocalDateTime만 다뤄서 LocalDateTime에는 필요하지 않은 정보(연도, 월, 일)이 있기에 제가 필요한 시간과 분만 가져오기 위해서 이렇게 따로 두었습니다!
좀 더 자세히 공부하고 적용하면 좋았을거 같네요!

Rok93 commented on src/main/java/dto/AbsenceCrewDto.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public int compareTo(AbsenceCrewDto o) {
+        int myAbsenceCount = results.getOrDefault(LATE.getResult(), 0);
+        myAbsenceCount += results.getOrDefault(ABSENCE.getResult(), 0) * LATE_TO_ABSENCE_THRESHOLD;
+        int otherAbsenceCount = o.results.getOrDefault(LATE.getResult(), 0);
+        otherAbsenceCount += o.results.getOrDefault(ABSENCE.getResult(), 0) * LATE_TO_ABSENCE_THRESHOLD;
```
아래의 요구사항에 따르면  잘 생각해보면 AttendanceResult의 각 요소별로 우선순위가 정해져있는데요. 
이 또한 도메인 로직으로 볼 수 있지 않을까요? 🤔 
이런 중요도를 AttendanceResult의 상태 값으로 가지고 이 값을 기준으로 정렬한다면 이 요구사항은 지금보다 더 깔끔하게 구현할 수 있지 않을까요? 😃 

> 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다

jaehyeon2650 replied:
> 제적 위험자 조회 결과
- 빙티: 결석 3회, 지각 4회 (면담)
- 이든: 결석 2회, 지각 5회 (면담)
- 빙봉: 결석 1회, 지각 6회 (면담)
- 쿠키: 결석 2회, 지각 3회 (면담)
- 짱수: 결석 0회, 지각 6회 (경고)

제적 위험자 조회 결과에 대한 예시입니다. 여기서 같은 대상 위험자는 지각을 결석으로 간주하여 내림차순하라는 요구사항이 있습니다. 지각과 결석을 동일하다고 하면 이든과 빙봉의 위치가 바뀌어야합니다. 근데 지금은 이든이 먼저 오는 것으로 보아 지각3 == 결석1로 계산하는게 맞다고 판단하였습니다. 그래서 결석을 지각 3회로 바뀌어서 지각이 많은 순서대로 정렬을 했습니다. 이렇게 로직을 구성하면 재적 위험 상태를 정렬은 자동으로 만족하게 되므로 현재처럼 정렬을 했습니다.

저는 정렬은 도메인 로직으로 보지 않고 단순히 뷰의 역할로 판단하여 dto에서 정렬을 하는 선택을 했습니다! 이런 정렬도 도메인 로직인가요?? 저는 도메인 로직을 제적 위험자 조회만은 바라봤습니다! 
로키는 어디까지 도메인 로직이라고 기준을 정하시는지 궁금합니다!!

Rok93 replied:
실제 정렬에 대한 것 또한 요구사항이라고 생각해요. 
또한 뷰 단에서 정렬하는 것에 대한 한계점(많은 양의 데이터를 페이징 단위로 조회하는 경우, 정렬 조건이 여러개인 경우 etc)도 많다고 생각해요.  따라서 저는 도메인 로직으로 보는게 맞다고 생각해요.

jaehyeon2650 replied:
아하 이렇게 생각하니 정렬도 도메인의 로직이라고 생각이 되네요! 도메인으로 옮겨보도록 하겠습니다!

Rok93 commented on src/main/java/util/CsvReader.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static Map<String, List<LocalDateTime>> loadAttendanceData() {
+        Map<String, List<LocalDateTime>> crewsMap = new HashMap<>();
+        try (Scanner scanner = new Scanner(new File(FILE))) {
```
try with resource 활용 👍💯

Rok93 commented on src/main/java/util/CsvReader.java:
```diff
@@ -0,0 +1,51 @@
+        try (Scanner scanner = new Scanner(new File(FILE))) {
+            readLines(scanner, crewsMap);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
```
[표준 예외를 사용](https://velog.io/@semi-cloud/Effective-Java-%EC%95%84%EC%9D%B4%ED%85%9C-72-%ED%91%9C%EC%A4%80-%EC%98%88%EC%99%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)해보면 어떨까요?

jaehyeon2650 replied:
표준 예외라는 개념은 처음 들어봤습니다! 해당 자료를 통해 공부한 후 적용해보도록 하겠습니다!

Rok93 reviewed (changes requested):
코기 안녕하세요. 출석 미션을 함께하게된 로키입니다. 
1단계 구현 잘해주셨습니다. 👍👍 

몇몇 코멘트 남겨두었는데, 확인해서 반영 부탁드릴게요. 🙏

jaehyeon2650 reviewed (commented):
객체를 구성하면서 역할이 애매한 부분이 존재하였고 해결하지 못하였는데 로키의 의견 덕분에 리팩토링 틀이 잡히네요! 소중한 리뷰 감사드립니다!😊
네이밍, 코드 가독성을 위한 정렬 등 놓친 부분도 알려주셔서 감사합니다!

jaehyeon2650 committed:
feat(domain): DailyAttendanceTime enum 타입 추가

- 각 요일별 출석 시간을 상수화
- 각 요일별 출석 결과를 도출하는 기능 추가
- 각 요일별 출석이 가능한지 도출하는 기능 추가

jaehyeon2650 committed:
test(domain): DailyAttendanceTime enum 테스트

- 각 요일별 출석 결과를 도출하는 기능 테스트
- 각 요일별 출석이 가능한지 도출하는 기능 테스트

jaehyeon2650 committed:
refactor : 코드 컨벤션 수정

jaehyeon2650 committed:
rename(domain): History -> AttendanceHistory 이름 변경

jaehyeon2650 committed:
rename(domain): AbsenceLevel -> AbsencePenalty 이름 변경

jaehyeon2650 committed:
refactor(controller): 비교자 변경

jaehyeon2650 committed:
refactor(utils): 표준 예외 변경

Rok93 commented on src/main/java/constants/SelectionOption.java:
```diff
@@ -0,0 +1,25 @@
+        return Arrays.stream(SelectionOption.values())
+                .filter(select -> select.option.equals(input))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("[ERROR] 잘못된 입력입니다."));
```
어떤 값이 들어왔는지도 예외 메시지로 보여주면 디버깅하기가 조금 더 수월할 것 같아요.

jaehyeon2650 replied:
넵 수정해보도록 하겠습니다!

Rok93 commented on src/main/java/constants/SelectionOption.java:
```diff
@@ -0,0 +1,25 @@
+        this.option = option;
+    }
+
+    public static SelectionOption getSelectOption(String input) {
```
메서드 파라미터로 들어온 값이 입력으로 들어온 값인지 알 필요가 있을까요? 🤔 
메서드 파라미터로 들어오는 값이 어떤 값인지 네이밍으로 드러내주면 좋을 것 같아요. 😃

jaehyeon2650 replied:
아하 그렇네요! InputView와 연관되게 생각해서 이렇게 네이밍을 지은거 같아요! 로키 말대로 어떻게 들어오는지가 아니라 어떤 값인지를 표현하는게 좋아보여요! 감사합니다

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,95 @@
+
+    private final OutputVIew outputVIew;
+    private final InputView inputView;
+    private final Crews crews;
```
Config 클래스쪽엔 남긴 (2번) 코멘트의 연장선인데, 비슷한 이유로 Controller에서 상태 값으로 Crews를 가지기 보다는 필요한 시점에 Crews 객체를 생성하도록해보면 좋을 것 같아요.

jaehyeon2650 replied:
네 수정해보도록 하겠습니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,95 @@
+            Map<String, Integer> results = crews.getAttendanceAllResult(member.getUserName(), newDate);
+            AbsencePenalty classifyAbsencePenalty = crews.getClassifyAbsenceLevel(member.getUserName(), newDate);
+            return new AbsenceCrewDto(member.getUserName(), results, classifyAbsencePenalty);
+        }).collect(Collectors.toList());
```
자바 16부터 `toList()` 함수를 제공해주니, 대신 활용해볼 수 있을 것 같아요. 😃

jaehyeon2650 replied:
아! 이 부분은 제가 DTO에서 제적 위험자에 대해서 정렬하고 있습니다. 그렇기에 toList() 함수를 사용하면 불변 리스트로 리턴하여 정렬 기능에서 예외가 발생합니다!
때문에 collect(Collectors.toList())로 변경 가능한 리스트를 만들었습니다! 로키 말대로 정렬하는 로직을 도메인으로 넣으면 해당 메서드는 toList() 함수로 바꾸면 될거 같습니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,95 @@
+            Map<String, Integer> results = crews.getAttendanceAllResult(member.getUserName(), newDate);
+            AbsencePenalty classifyAbsencePenalty = crews.getClassifyAbsenceLevel(member.getUserName(), newDate);
+            return new AbsenceCrewDto(member.getUserName(), results, classifyAbsencePenalty);
+        }).collect(Collectors.toList());
```
이 부분은 crews [객체에게 메시지를 보내서](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/) 결과만 받아볼 수 없을까요? 🤔

받아온 결과에서 필요한 값들만 추출하여 DTO로 변환해주면 될 것 같아서요. 🤔

Rok93 commented on src/main/java/domain/Histories.java:
```diff
@@ -0,0 +1,108 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class Histories {
```
AttendanceHistory의 일급컬렉션이니 `AttendanceHistories`로 네이밍을 변경해주면 어떨까요? 😃

jaehyeon2650 replied:
네 좋습니다!

Rok93 commented on src/main/java/domain/AbsencePenalty.java:
```diff
@@ -0,0 +1,35 @@
+    NORMAL("정상", 0);
+
+    public static final int LATE_TO_ABSENCE_THRESHOLD = 3;
+    private final String level;
```
상수와 상태를 구분하기 쉽도록 절적히 줄바꿈을 통해 구분해주세요. 😃

```suggestion
    public static final int LATE_TO_ABSENCE_THRESHOLD = 3;
    
    private final String level;
```

jaehyeon2650 replied:
앗 알겠습니다!

Rok93 commented on src/main/java/domain/AbsencePenalty.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static AbsencePenalty findAbsenceLevel(int absentCount, int lateCount) {
+        absentCount += (lateCount / LATE_TO_ABSENCE_THRESHOLD);
```
받은 메서드 파라미터를 수정하기보다는 필요하다면 별도의 변수를 둬서 사용하는 것이 좋다고생각해요. 

```suggestion
    public static AbsencePenalty findAbsenceLevel(final int absentCount, final int lateCount) {
        int calculatedAbsentCount = absentCount + (lateCount / LATE_TO_ABSENCE_THRESHOLD);
```

jaehyeon2650 replied:
좋은 의견 감사합니다!😊

Rok93 commented on src/main/java/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,34 @@
+
+    public String getAttendanceResult() {
+        return attendanceResult.getResult();
+    }
```
가능하면 getter, setter 를 가장 최하단에 위치시키면 좋을 것 같아요. 
다른 객체와 협력하기위한 책임들이 더 우리가 관심있게 봐야할 대상 메서드라고 생각하기 때문이에요. 😃 (getter setter를 예시로 들었지만, 관용적인 메서드(ex. compareTo)들도 가능한 아래에 위치시키는게 좋다고 생각해요. 🤔

jaehyeon2650 replied:
메서드에 위치에 대해서는 신경을 쓰지 않았는데 로키 말을 들어보면 메서드 위치도 중요하다고 생각되네요!
좋은 의견 감사드립니다 🥹

Rok93 commented on src/main/java/domain/DailyAttendanceTime.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static void validateOpeningHours(LocalDateTime attendanceTime) {
+        if (!(attendanceTime.getHour() == ABSENT_DEFAULT_HOUR && attendanceTime.getMinute() == ABSENT_DEFAULT_MINUTE)
```
23시 59분에만 예외처리를 하는건가요? 
예를들어 23시 01분에 출석을하면 어떻게 되는건가요? 

이 조건식이 조금 파악하기가 어려운 것 같은데요. 😅
의미있는 단위로 메서드로 분리해주면 조금 더 가독성이 좋아질 것 같아요.

jaehyeon2650 replied:
아 이 부분에 대해 리드미에 정확히 작성하지 않았네요!
페어와 함께 특정 날짜에 출석 기록이 없는 경우 즉, 아예 결석을 한 경우는 어떻게 처리할지 고민을 하였습니다. 8시부터 23시까지는 출석이 불가능하니 출석 자체를 안한 경우를 임의로 23시 59분이라고 정했습니다!
정확한 네이밍과 리드미 작성으로 이해하기 쉽도록 변경해야할거같습니다!

Rok93 replied:
음... 이 부분은 조금 고민되는 것 같아요. 
> 8시부터 23시까지는 출석이 불가능하니 출석 자체를 안한 경우를 임의로 23시 59분이라고 정했습니다!

출석 자체를 안했다면 출석 시간이 없는 것인데 임의의 시간을 넣어주게되면 오히려 혼동이 생기지 않을까요? 🤔 
0 ~ 8시 와 23시 ~ 24시 사이에 출석을 한 경우에는 출석을 남기되 결석처리를 한다거나 혹은 애초에 출석을 할 수 없도록 만든다거나 등의 방식이 조금 덜 혼동스럽지 않을까싶어요. 

당장 출석시간이 `8시 ~ 24시`로 변경된다면 미출석자들은 어떻게 될까요? 🤔

jaehyeon2650 replied:
맞습니다! 그 부분을 설계하면서 어려웠던 부분입니다. 그래도 현재를 기준으로 23시 59분 59초는 안쓴다고 판단하여 이를 결석용으로 사용했는데 로키 말대로 출석 시간이 변경되면 문제가 될거 같습니다.
step2에서는 다른 방법을 생각하고 적용해보도록 하겠습니다.

Rok93 commented on src/main/java/domain/Histories.java:
```diff
@@ -0,0 +1,108 @@
+        this.histories = copy.stream().map(AttendanceHistory::new).collect(Collectors.toList());
+    }
+
+    public Map<String, Integer> getAttendanceResultCount(LocalDateTime standard) {
```
가능한 `getXXX`와 같은 네이밍보다는 조금 더 행위를 드러낼 수 있는 네이밍을 사용해보시면 좋을 것 같아요. 😃

지금의 경우에는 저라면 `countAttendanceResult`라고 명명했을 것 같아요. 

[이 글](https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/)을 참고해보시면 도움이되실 것 같아요. 😉

jaehyeon2650 replied:
네이밍이 하는건 여전히 어려운거 같습니다..! ㅠㅠ 해당 자료를 참고하여 수정하도록 하겠습니다!

Rok93 commented on src/main/java/view/OutputVIew.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Map;
+import util.Convertor;
+
+public class OutputVIew {
```
```suggestion
public class OutputView {
```

I는 소문자여야할 것 같네요. 👀

jaehyeon2650 replied:
앗..! 사소한 실수네요 ㅜ 지적 감사합니다

Rok93 commented on src/test/java/domain/HistoriesTest.java:
```diff
@@ -0,0 +1,262 @@
+
+    @Test
+    @DisplayName("기록 삭제 테스트 - 정상")
+    void deleteHistory() {
```
가능한 테스트는 실제 메서드 네이밍과 구분이되도록  prefix로 test를 붙여주거나해주면 좋을 것 같아요. 🤔

jaehyeon2650 replied:
넵 알겠습니다! 테스트 이름도 수정해보도록 하겠습니다.

Rok93 commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,66 @@
+        return findCrew.getBeforeHistories(standard);
+    }
+
+    public void addHistory(String username, LocalDateTime attendanceTime) {
```
실제 현실 세계에서의 출석을 기준으로 생각해봤을 때,  이 `Attendance`를 추가하는 행위를 뭐라고할까요? 😃
조금 더 현실에 빗대어서 네이밍을 지어준다면 더 직관적으로 코드를 이해하기가 좋을 것 같아요. 😃

jaehyeon2650 replied:
넵! 네이밍 충분히 고민하고 바꿔보도록 하겠습니다.

Rok93 commented on src/main/java/domain/DailyAttendanceTime.java:
```diff
@@ -0,0 +1,80 @@
+                                                        LocalDateTime attendanceTime) {
+        LocalTime todayAttendanceTime = dailyAttendanceTime.attendanceTime;
+        LocalTime attendanceLocalTime = attendanceTime.toLocalTime();
+        if (attendanceLocalTime.isAfter(todayAttendanceTime.plusMinutes(ABSENCE_STANDARD_MINUTES))) {
```
사소하지만 Duration을 활용해서 분을 체크하는 방식도 가능하겠네요. 😃 (이 부분은 꼭 수정해달라는 것이 아니고 Duration을 활용하는 방식도 있다 정도로 생각해주시면 될 것 같아요)

```suggestion
        if (Duration.between(attendanceLocalTime, todayAttendanceTime).toMinutes() > ABSENCE_STANDARD_MINUTES) {
```

jaehyeon2650 replied:
헉! Duration은 처음봅니다! 좋은 정보 감사합니다! 수정하면서 공부해보고 싶네요 ㅎㅎ

Rok93 commented on src/main/java/domain/AbsencePenalty.java:
```diff
@@ -0,0 +1,35 @@
+
+    public static AbsencePenalty findAbsenceLevel(int absentCount, int lateCount) {
+        absentCount += (lateCount / LATE_TO_ABSENCE_THRESHOLD);
+        if (absentCount > OUT.standard) {
+            return OUT;
+        }
+        if (absentCount >= MEETING.standard) {
+            return MEETING;
+        }
+        if (absentCount >= WARNING.standard) {
+            return WARNING;
+        }
+        return NORMAL;
```
추후 enum 요소가 추가될 때마다 해당 로직도 수정될 것 같은 느낌이드는데요. 
enum 요소가 추가 혹은 삭제되더라도 이 메서드를 수정하지 않는 방법은 없을까요? 😃

jaehyeon2650 replied:
스트림을 활용해서 if문을 사용하지 않고 해당 로직과 동일하게 실행될 방법을 고민해보겠습니다!

Rok93 reviewed (changes requested):
코기 피드백 반영 잘해주셨어요! 👍👍
조금 더 짚고 넘어가면 좋을 부분이보여서 추가 코멘트 남겨두었습니다. 😃
추가적으로 남겨주신 질문에 대한 답변들도 추가해뒀으니 같이 확인 부탁드릴게요. (혹시 질문에 누락된 답변이 있다면 말씀해주세요)

jaehyeon2650 reviewed (commented):
의문점에 대해서 로키의 의견을 정확히 말해주셔서 이해하기 쉬웠습니다! 덕분에 생각하는 범위가 넓어지는거 같습니다! 감사합니다 😊

jaehyeon2650 committed:
refactor(configure): Configure 리펙토링

- 컨트롤러 Crews 맴버 변수 삭제
- 컨트롤러 시작시 Crews 객체를 초기화하도록 변경
- Configure 클래스에서 초기화하는 로직 삭제

jaehyeon2650 committed:
refactor(constants): SelectionOption 리펙토링

- 메서드 파라미터 네이밍 수정
- 예외 메시지 수정

jaehyeon2650 committed:
refactor(AbsencePenalty): AbsencePenalty 리펙토링

- findAbsenceLevel 메서드 로직 변경
- findAbsenceLevel 메서드 로직 테스트

jaehyeon2650 committed:
refactor(AttendanceHistory): AttendanceHistory 리펙토링

- 메서드 위치 변경(getter 메서드 하단으로 이동)
- 맴버 변수 위치 변경

jaehyeon2650 committed:
refactor(Crews): Crews 리펙토링

- 메서드 이름 변경

jaehyeon2650 committed:
refactor(DailyAttendanceTime): DailyAttendanceTime 리펙토링

- 출석 결과 판단하는 로직 변경

jaehyeon2650 committed:
refactor(AttendanceHistories): AttendanceHistories 리펙토링

- 클래스 이름 변경 Histories -> AttendanceHistories
- 메서드 이름 변경
- 테스트 이름 변경

jaehyeon2650 committed:
refactor(OutputView): OutputView 클래스 이름 수정

jaehyeon2650 committed:
refactor(test): 테스트 메서드 이름 변경

jaehyeon2650 committed:
fix(DailyAttendanceTime): 지각 계산 로직 버그 수정

jaehyeon2650 committed:
refactor: 출석 기록 로직 변경

- AttendanceRecord 도메인 생성
- AbsenceCrewDto 정렬 기능 삭제
- 제적 위험자 정렬 기능 도메인 내부로 이동
- 출석 기록 관련 DTO 생성자 변경

jaehyeon2650 committed:
refactor(domain) : 크루 compareTo 메서드 추출

jaehyeon2650 committed:
refactor(test) : AttendanceRecord 관련 테스트 버그 수정

jaehyeon2650 committed:
test(AttendanceRecord) : AttendanceRecord 도메인 테스트

jaehyeon2650 committed:
refactor: 코드 정리

jaehyeon2650 committed:
refactor(crews): 메서드명 변경

jaehyeon2650 committed:
refactor(domain): AttendanceRecord 메서드 위치 변경

Rok93 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,34 @@
+    CHRISTMAS(12, 25),
+    NEW_YEARS_DAY(1, 1);
+
+    private final int month;
+    private final int day;
```
저도 잘 사용하지 않아서 이런게 있는지는 몰랐었는데, 역시 찾아보니 있네요. 
월,일 정보만 필요하다면 LocalDate 대신에 [MonthDay](https://lts0606.tistory.com/257)를 활용해보면 어떨까요?

Rok93 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,34 @@
+        this.day = day;
+    }
+
+    public static void validateHoliday(LocalDateTime time) {
```
Holiday 객체 내부에서 등교일에 관련된 예외를 던지는게 조금 어색하게 느껴지는데요 🤔

`DailyAttendanceTime.getDailyAttendanceResult` 내부에서 isHoliday 인지 확인해서 예외를 던지는게 조금 더 자연스러운 흐름이될 것 같아요. 😃

jaehyeon2650 replied:
Holiday 객체 자체는 공휴일인지를 판단하는 클래스인데 로키 말대로 예외를 던지는 것은 어색해보입니다! 수정해보도록 하겠습니다.

Rok93 commented on src/main/java/dto/AbsenceCrewsDto.java:
```diff
@@ -0,0 +1,29 @@
+public class AbsenceCrewsDto {
+    private final List<AbsenceCrewDto> crews;
+
+    public AbsenceCrewsDto(Map<Crew, AttendanceRecord> record) {
```
DTO의 경우 [Record](https://colevelup.tistory.com/28)를 활용해도 좋을 것 같네요. 😃 

지금보니 HistoriesDto는 record 클래스를 활용하셨군요? 
여기서는 일반 클래스를 선언하신 이유가 있으실까요?

jaehyeon2650 replied:
따로 이유는 없었습니다..! 페어 프로그래밍하면서 꼬인거 같습니다! record 클래스로 통일하는게 좋아보입니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private void editAttendance(Crews crews) {
+        String editName = inputView.getEditName();
+        LocalDateTime editHistory = inputView.getEditAttendanceTime();
```
출석 확인, 출석 수정 등을 할 때, 없는 크루의 닉네임을 입력하면, 다음 입력을 하기 전에 `[ERROR]  등록되지 않은 닉네임입니다.` 에러 메시지가 출력되어야할 것 같아요. 🤔 

<img width="1002" alt="image" src="https://github.com/user-attachments/assets/ec85f50e-cbfc-4ece-b7d6-2d1106b69f4b" />

<img width="390" alt="image" src="https://github.com/user-attachments/assets/771f0e51-8655-45b7-ae01-456792e51553" />

jaehyeon2650 replied:
step2에서는 주의깊게 작성하도록 하겠습니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private void editAttendance(Crews crews) {
+        String editName = inputView.getEditName();
+        LocalDateTime editHistory = inputView.getEditAttendanceTime();
```
출석일과 출석 시간을 `LocalDateTime` 으로 합치셨군요? 
이 부분은 출석일자(yyyy-mm-dd)와 출석시간(hh:mm:ss)을 분리한채로 두는게 더 낫지 않을까요? 🤔

jaehyeon2650 replied:
오 좋은 생각입니다! 출석 시간만을 뽑으려고 하는 경우가 많아서 애초에 분리해놓는게 좋아보입니다!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,93 @@
+        String editName = inputView.getEditName();
+        LocalDateTime editHistory = inputView.getEditAttendanceTime();
+        LocalDateTime beforeHistory = crews.getAttendanceHistory(editName, editHistory);
+        String beforeResult = crews.getRecordedAttendanceResult(editName, editHistory);
+        String editResult = crews.editAttendanceHistory(editName, editHistory);
```
getRecordedAttendanceResult, editAttendanceHistory의 결과는  `AttendanceResult` 타입을 반환하되, print를 할 때에는 getResult를 호출하여 View에 String을 전달해주면 좋을 것 같아요.

Rok93 commented on src/main/java/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,35 @@
+
+public class AttendanceHistory implements Comparable<AttendanceHistory> {
+    public final static int ABSENT_DEFAULT_HOUR = 23;
+    public final static int ABSENT_DEFAULT_MINUTE = 59;
+    private final LocalDateTime attendanceTime;
```
상수와 상태는 줄바꿈을 통해서 적절히 분리해주세요. 😃 
```suggestion
    public final static int ABSENT_DEFAULT_MINUTE = 59;
    
    private final LocalDateTime attendanceTime;
```

Rok93 commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public Map<AttendanceResult, Integer> getTotalAttendanceRecord() {
+        Map<AttendanceResult, Integer> result = new EnumMap<>(AttendanceResult.class);
+        result.put(ATTENDANCE, attendanceCount);
+        result.put(LATE, lateCount);
+        result.put(ABSENCE, absenceCount);
+        return result;
```
아래와 같이 [컬렉션 팩토리](https://devbksheen.tistory.com/entry/%EB%AA%A8%EB%8D%98-%EC%9E%90%EB%B0%94-%EC%BB%AC%EB%A0%89%EC%85%98-%ED%8C%A9%ED%86%A0%EB%A6%ACCollection-Factory-%EC%86%8C%EA%B0%9C-%EB%B0%8F-%EC%82%AC%EC%9A%A9%EB%B2%95)를 활용해볼 수 있을 것 같아요. 

```suggestion
        return Map.of(
            ATTENDANCE, attendanceCount,
            LATE, lateCount,
            ABSENCE, absenceCount
        );
```

jaehyeon2650 replied:
확실히 깔끔하네요! 개선된 방법 알려주셔서 감사합니다 🥹

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,95 @@
+        this.attendanceHistories = new AttendanceHistories(histories, standard);
+    }
+
+    public void validateName(String userName) {
```
이 메서드는 Crew 내부에서만 사용되는 것 같은데요. 
접근 제한자를 private로 제한해주면 좋을 것 같아요. 😃

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    public void validateName(String userName) {
+        if (userName.length() >= 5) {
```
5라는 숫자는 다른 사람 입장에서는 한눈에 이해하기 어려운 숫자인 것 같아요. 👀
적절히 상수로 추출하여 의도를 드러내보면 어떨까요? 😃

추가적으로  이름이 네 글자 이하여야한다고 했으면,  5보다는 4가 조금 더 의도를 드러내기가 쉬운 숫자일 것 같아요. 
```suggestion
        if (userName.length() > 4) {
```

jaehyeon2650 replied:
앗 놓친 부분이네요! 로키 말대로 상수로 추출하여 의도를 정확히 드러내는 것이 좋아보입니다!

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    public int compareTo(Crew o) {
+        LocalDateTime standard = Convertor.changeStandardLocalDateTime(LocalDateTime.now());
+        Map<AttendanceResult, Integer> thisAttendanceResult = getAttendanceResultMap(standard, this);
+        Map<AttendanceResult, Integer> otherAttendanceResult = getAttendanceResultMap(standard, o);
+        int result = compareAttendanceResult(thisAttendanceResult, otherAttendanceResult);
+        if (result == 0) {
+            return this.userName.compareTo(o.userName);
+        }
+        return result;
```
이건 정렬 순서가 어떻게 되는걸까요?

jaehyeon2650 replied:
결석 1회 -> 지각 3회로 바꾸어서 지각이 가장 많은 사람 순으로 정렬했습니다. 같은 경우 이름 순으로 정렬됩니다. 이렇게 하면 위험자 관련 정렬을 한번에 해결할 수 있어서 선택했습니다.

Rok93 reviewed (approved):
코기 피드백 반영 잘해주셨네요. 👍👍 
몇몇 코멘트 남겨두었는데, 다음 단계 진행하기 전에, 먼저 반영해주시고, 
이어서 다음 단계 진행해주시면 될 것 같습니다. 😃
다음 단계에서 뵐게요. 🙏

jaehyeon2650 reviewed (commented):
시간내어 리뷰 남겨주셔서 감사합니다! 해당 코멘트는 Step2에서 반영하도록 하겠습니다!
